### PR TITLE
Formalize the ProgressReporting Type

### PR DESCRIPTION
**Pull request by soulomoon**
follow up of https://github.com/haskell/haskell-language-server/pull/4205
This is intended to remove the useless `inProgress` when tracking the `inProgressState` from the outside.
1. Add kind `data ProgressStateTrackingType = ProgressStateInsideTracking | ProgressStateOutsideTracking` to distinguish the two type of `inProgressState` tracking.
2. Spread the `TProgressState` to related type location to mark what kind of tracking we used. Then we won't have `inProgress` for `ProgressStateOutsideTracking`

### CODE REVIEW

**Comment by michaelpj** [comment]
I'm not sure I see the attraction of this design, to be honest. I think we could instead see the tracking with files as a _wrapper_ on top of the simpler `progressCounter` that just works with the `STM Int`s. That would mean we have two different types for the two different counters, rather than just one with a type parameter, but I think it would be simpler?

**Comment by soulomoon** [comment]
Good suggestions @michaelpj , 
I've managed to use the wrapper version,
Seems much simpler using the wrapper version here.

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs [hunk 1]
```diff
@@ -42,38 +48,60 @@ data ProgressEvent
   | ProgressCompleted
   | ProgressStarted
 
-data ProgressReporting m = ProgressReporting
-  { progressUpdate :: ProgressEvent -> m (),
-    inProgress     :: forall a. NormalizedFilePath -> m a -> m a,
-    -- ^ see Note [ProgressReporting API and InProgressState]
-    progressStop   :: IO ()
+data ProgressReportingNoTrace m = ProgressReportingNoTrace
+  { progressUpdateI :: ProgressEvent -> m (),
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/ProgressReporting.hs:52]: You might not need to add name suffixes if you use `DuplicateRecordFields`?

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs [hunk 2]
```diff
@@ -42,38 +48,60 @@ data ProgressEvent
...
   }
 
+data ProgressReporting m = ProgressReporting
+  {
+    inProgress             :: forall a. NormalizedFilePath -> m a -> m a,
+    -- ^ see Note [ProgressReporting API and InProgressState]
+    progressReportingInner :: ProgressReportingNoTrace m
+  }
+
+class ProgressReportingClass a where
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/ProgressReporting.hs:65]: Maybe just `ProgressReporter`? Usually we read class constraints as "X is a Y" or similar, so "X is a ProgressReporter" reads sensibly?

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs [hunk 3]
```diff
@@ -42,38 +48,60 @@ data ProgressEvent
...
+    type M a :: * -> *
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/ProgressReporting.hs:66]: I think I asked this before, but can everything just be in IO? The only place we use another monad is in `inProgress`, and I think we could also just provide that with IO and get people to use `MonadUnliftIO` to use it?

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs [hunk 4]
```diff
@@ -42,38 +48,60 @@ data ProgressEvent
...
-    -- ^ see Note [ProgressReporting API and InProgressState]
-    progressStop   :: IO ()
+data ProgressReportingNoTrace m = ProgressReportingNoTrace
+  { progressUpdateI :: ProgressEvent -> m (),
+    progressStopI   :: IO ()
     -- ^ we are using IO here because creating and stopping the `ProgressReporting`
     -- is different from how we use it.
   }
 
+data ProgressReporting m = ProgressReporting
```
Comment (michaelpj) [inline ghcide/src/Development/IDE/Core/ProgressReporting.hs:58]: I would name the other one `ProgressReporting` and call this `PerFileProgressReporting` or something?

**Comment by michaelpj** [comment]
Looks much better

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs (hunk 3 above)
Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/ProgressReporting.hs:66]: `inProgress`'s monad `m` is invariant (Both covariant and contra variant locations). I doubt we can have `forall a.IO a -> IO a` -> `forall a. m a -> m a`.🤔

Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/ProgressReporting.hs:66]: Yes, we can turn them into IOs and save the `m`

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs (hunk 1 above)
Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/ProgressReporting.hs]: No, we have class function and data field collision. `DuplicateRecordFields ` is not helpful here

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs (hunk 2 above)
Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/ProgressReporting.hs]: good name

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs (hunk 4 above)
Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/ProgressReporting.hs]: make sense, I like this naming

File: ghcide/src/Development/IDE/Core/ProgressReporting.hs (hunk 1 above)
Comment (soulomoon) [reply ghcide/src/Development/IDE/Core/ProgressReporting.hs]: But maybe we can make it prefix with '_' and not export it.

**Review by michaelpj — APPROVED** [review]
Great

**Pull request merged**